calyx-backend/src/xilinx/idl.rs

```diff
@@ -0,0 +1,148 @@
+            .unwrap();
+
+        let memory_names = external_memories(toplevel);
+        let mem_infos = get_mem_info(toplevel);
+
+        let memories: Vec<Memory> = memory_names
+            .iter()
+            .zip(mem_infos.iter())
+            .map(|(memory_name, mem_info)| Memory {
+                name: memory_name,
+                width: mem_info.width,
+                size: mem_info.size,
+            })
+            .collect();
+
+        let program_interface = ProgramInterface {
+            toplevel_name: toplevel.name.as_ref(),
+            memories,
+        };
+
+        //TODO: we want to have na array of component names
+        // check how to get them
+
+        write!(
+            file.get_write(),
+            "{}",
+            serde_json::to_string_pretty(&program_interface)
+                .expect("IDL Serialization failed")
+        )?;
+
+        Ok(())
+    }
+}
+
+/// Parameters for single dimensional memory
+struct MemInfo {
+    width: u64,
+    size: u64,
+}
+
+// Gets all external memory cells in top level
+fn external_memories_cells(comp: &ir::Component) -> Vec<ir::RRC<ir::Cell>> {
+    comp.cells
+        .iter()
+        // find external memories
+        .filter(|cell_ref| {
+            let cell = cell_ref.borrow();
+            cell.attributes.has(ir::BoolAttr::External)
+        })
+        .cloned()
+        .collect()
+}
+
+fn get_mem_info(comp: &ir::Component) -> Vec<MemInfo> {
+    external_memories_cells(comp)
+        .iter()
+        .map(|cr| {
+            let cell = cr.borrow();
+            let cell_size = match cell.prototype.get_name().unwrap().as_ref() {
+                "std_mem_d1" | "seq_mem_d1" => {
+                    cell.get_parameter("SIZE").unwrap()
+                }
+                "std_mem_d2" | "seq_mem_d2" => {
+                    cell.get_parameter("D0_SIZE").unwrap()
+                        * cell.get_parameter("D1_SIZE").unwrap()
+                }
+                "std_mem_d3" | "seq_mem_d3" => {
+                    cell.get_parameter("D0_SIZE").unwrap()
+                        * cell.get_parameter("D1_SIZE").unwrap()
+                        * cell.get_parameter("D2_SIZE").unwrap()
+                }
+
+                "std_mem_d4" | "seq_mem_d4" => {
+                    cell.get_parameter("D0_SIZE").unwrap()
+                        * cell.get_parameter("D1_SIZE").unwrap()
+                        * cell.get_parameter("D2_SIZE").unwrap()
+                        * cell.get_parameter("D3_SIZE").unwrap()
+                }
+                _ => {
+                    panic!("cell `{}' marked with `@external' but is not a memory primitive.", cell.name())
+                }
+            };
+
+            MemInfo {
+                width: cell.get_parameter("WIDTH").unwrap(),
+                size: cell_size,
+            }
+        })
+        .collect()
+}
+
+// Returns Vec<String> of memory names
+fn external_memories(comp: &ir::Component) -> Vec<String> {
```

Comment:
It looks like `external_memories` is identical to the definition in `top-level.rs`. (And `get_mem_info` is similar but does the "size flattening" thing internally?) What do you think about moving this stuff to a common utility file so both backends can use it, to avoid the duplication of code?

Reply:
Hopefully this is better now? @sampsyo for visibility.